beacon-chain/core/helpers/committee.go

```diff
@@ -395,3 +396,69 @@ func VerifyAttestationBitfield(bState *pb.BeaconState, att *pb.Attestation) (boo
 	}
 	return VerifyBitfield(att.AggregationBits, len(committee))
 }
+
+// CompactCommitteesRoot returns the index root of a given epoch.
+//
+// Spec pseudocode definition:
+//   def get_compact_committees_root(state: BeaconState, epoch: Epoch) -> Hash:
+//    """
+//    Return the compact committee root for the current epoch.
+//    """
+//    committees = [CompactCommittee() for _ in range(SHARD_COUNT)]
+//    start_shard = get_epoch_start_shard(state, epoch)
+//    for committee_number in range(get_epoch_committee_count(state, epoch)):
+//        shard = Shard((start_shard + committee_number) % SHARD_COUNT)
+//        for index in get_crosslink_committee(state, epoch, shard):
+//            validator = state.validators[index]
+//            committees[shard].pubkeys.append(validator.pubkey)
+//            compact_balance = validator.effective_balance // EFFECTIVE_BALANCE_INCREMENT
+//            # `index` (top 6 bytes) + `slashed` (16th bit) + `compact_balance` (bottom 15 bits)
+//            compact_validator = uint64((index << 16) + (validator.slashed << 15) + compact_balance)
+//            committees[shard].compact_validators.append(compact_validator)
+//    return hash_tree_root(Vector[CompactCommittee, SHARD_COUNT](committees))
+func CompactCommitteesRoot(state *pb.BeaconState, epoch uint64) ([32]byte, error) {
+	shardCount := params.BeaconConfig().ShardCount
+	compactCommList := make([]*pb.CompactCommittee, shardCount)
+	for i := range compactCommList {
+		compactCommList[i] = &pb.CompactCommittee{}
+	}
+	comCount, err := CommitteeCount(state, epoch)
+	if err != nil {
+		return [32]byte{}, err
+	}
+	startShard, err := StartShard(state, epoch)
+	if err != nil {
+		return [32]byte{}, err
+	}
+	for i := uint64(0); i < comCount; i++ {
+		shard := (startShard + i) % shardCount
+		crossComm, err := CrosslinkCommittee(state, epoch, shard)
+		if err != nil {
+			return [32]byte{}, err
+		}
+
+		for _, indice := range crossComm {
+			validator := state.Validators[indice]
+			compactCommList[shard].Pubkeys = append(compactCommList[shard].Pubkeys, validator.Pubkey)
+			compactValidator := compressValidator(validator, indice)
+			compactCommList[shard].CompactValidators = append(compactCommList[shard].CompactValidators, compactValidator)
+
+		}
+	}
+	return ssz.HashTreeRoot(compactCommList)
+}
+
+func compressValidator(validator *pb.Validator, idx uint64) uint64 {
```

Comment:
can you add a function comment? what is it used for?.. etc

Reply:
done